src/engine/MaxConstraint.h

```diff
@@ -34,137 +46,197 @@ class MaxConstraint : public PiecewiseLinearConstraint
+
+    /*
+       Checks whether a case has been flagged as infeasible. Works only in
+       context-dependent mode.
+     */
+    bool isCaseInfeasible( unsigned phase ) const;
+
+    /*
+      Check if the constraint's phase has been fixed.
     */
-    PiecewiseLinearCaseSplit getValidCaseSplit() const;
+    bool phaseFixed() const override;
 
     /*
       Return a list of smart fixes for violated constraint.
     */
-    List<PiecewiseLinearConstraint::Fix> getSmartFixes( ITableau *tableau ) const;
+    List<PiecewiseLinearConstraint::Fix> getSmartFixes( ITableau *tableau ) const override;
 
     /*
       Preprocessing related functions, to inform that a variable has been eliminated completely
       because it was fixed to some value, or that a variable's index has changed (e.g., x4 is now
       called x2). constraintObsolete() returns true iff and the constraint has become obsolote
       as a result of variable eliminations.
     */
-    void eliminateVariable( unsigned variable, double fixedValue );
-    void updateVariableIndex( unsigned oldIndex, unsigned newIndex );
-    bool constraintObsolete() const;
+    void eliminateVariable( unsigned variable, double fixedValue ) override;
+    void updateVariableIndex( unsigned oldIndex, unsigned newIndex ) override;
+    bool constraintObsolete() const override;
 
     /*
       Get the tightenings entailed by the constraint.
     */
-    void getEntailedTightenings( List<Tightening> &tightenings ) const;
+    void getEntailedTightenings( List<Tightening> &tightenings ) const override;
 
     /*
       For preprocessing: get any auxiliary equations that this constraint would
       like to add to the equation pool.
     */
-    void addAuxiliaryEquations( InputQuery &inputQuery );
+    void addAuxiliaryEquations( InputQuery &inputQuery ) override;
 
     /*
       Returns string with shape:
       max, _f, element_1, element_2, ... , element_n
     */
-    String serializeToString() const;
+    String serializeToString() const override;
 
     /*
      * Returns a boolean value indicating if at least one input variable was eliminated (True)
      * or not (False)
      */
     bool wereVariablesEliminated() const;
 
+
+
+    bool isImplication() const override;
+
 private:
     unsigned _f;
     Set<unsigned> _elements;
     Set<unsigned> _initialElements;
 
-    unsigned _maxIndex;
-    bool _maxIndexSet;
     double _maxLowerBound;
     bool _obsolete;
     bool _eliminatedVariables;
     double _maxValueOfEliminated;
 
+    /*
+       Functions that abstract away _maxIndex and _maxIndexSet, and use the
+       refactored PhaseStatus to store this information.
+
+       Makes two assumptions:
+        - 0 is not a valid variable value, since it's reserved for PHASE_NOT_FIXED
+        - the total number of variables does not reach the value of MAX_PHASE_ELIMINATED
+     */
+    inline bool maxIndexSet() const
+    {
+        return getPhaseStatus() != PHASE_NOT_FIXED;
```

Comment:
I realized I don't fully understand the semantics of _phaseStatus and _cdPhaseStatus. Is this an indication of the phase that a non-fixed constraint is in? For MAX this makes sense (indicates the current element with maximal assignment), but what does it mean for Relu and Abs?

Reply:
In my mind the semantics is the same, the only difference is when we can act on the information. ReLU and other two-phase constraints do not depend on any additional variables, so we can treat them as implications immediately. Max is trickier because we need to convince ourselves that no other variable can exceed the value, before we can treat it as an implication.

Reply:
Sorry, I still don't follow. If max's phaseStatus() is x1, does this mean that currently x1 has the highest assignment?
Are we "overloading" this field to indicate this?

Reply:
Yes, x1 would have the highest assignment, or higher lower bound than all the other upper bounfs. 

I don't think the field is overloaded. 
Relu = max ( x, 0) vs Max = max( x, y). 

Relu can propagate as soon as we know something about x since zero is a constant. But max(x,y) can propagate only once it also knows that the upper bound of y is less than the value of x, otherwise it is not sufficient information. 

Basically, _phaseStatus can be thought of as the phase under the current bounds (and implicitly assignment). Relu needs information over one variable for propagation, but max needs also to know bounds on all the other variables involved to propagate. _phaseStatus denotes the same, the difference is that ReLU never has to wait on other variables so there is no period of uncertainty. Does this make sense?